### PR TITLE
Unsynced game commands cleanup (and others)

### PR DESCRIPTION
**Pull request by Svalorzen**
Given my last two pull requests, I'm just going through the executor's code and trying to cleaning them up and uniforming the way they parse arguments and so on.

I'm trying to keep each commit to each class so hopefully it's easier to review and fix/cherrypick/etc if needed. A couple commits sneaked in about another couple things, but I can remove them/move them to another PR if preferred.

I'm trying very hard to keep behavior intact, but obviously we should probably test these. I'm wondering if there's anything setup already for this.

I'm still not finished but I thought I'd open the PR anyway to let people see what I'm doing, if it makes any sense, or maybe I should just stop :)

This is currently based on #259 (since that's where I started making the fixes), but does not depend on it and I can rebase it on BAR105 directly if needed. I'm not sure if I can open a PR based on a PR and I'm doing something wrong, please do let me know if this is so.

### CODE REVIEW

**Comment by Beherith** [comment]
You can add to a PR by pushing to the same PR branch :) 

Also, I am _not_ going to test each changed action executor, so I would like to request that you do so :D 

Do tell when you are ready for merge!

**Comment by Svalorzen** [comment]
Eheheh that makes sense, I do expect to have to test it all. My main problem is that I'm not even sure what all these commands do. Is there a place that documents them so I can test my changes don't break anything?

I can read the code and try to guess the parameters but knowing exactly what is supposed to happen would help.

**Comment by Svalorzen** [comment]
Ah I do plan to add to this PR by pushing, I just wanted to have it out here in the meantime :)

**Comment by Svalorzen** [comment]
I'm around 1/3 done!

**Comment by Beherith** [comment]
> Eheheh that makes sense, I do expect to have to test it all. My main problem is that I'm not even sure what all these commands do. Is there a place that documents them so I can test my changes don't break anything?
> 
> I can read the code and try to guess the parameters but knowing exactly what is supposed to happen would help.

Which ones do you have questions about?

**Comment by Svalorzen** [comment]
I will make a list of all the commands I've cleaned up in the end and as I'm testing them I'll also ask if there's something I'm not sure what it's supposed to do :)

In the meantime, there's something I'd like to fix but I haven't yet because it impacts the Lua interface. In particular, the `DistIconActionExecutor` and some other executors read and `Set` integer parameters as arguments, but actually use them as float parameters (and cast them appropriately). Having them `float` in the fist place I think would make more sense, but as far as I could see it would require updating some Lua calls from `GetConfigInt` to `GetConfigFloat`, both here and in the `BAR` repository (doesn't look like there are many calls for these params). How is a change like this synced between the two repos?

**Comment by Svalorzen** [comment]
I haven't yet changed it, but I can say with 100% confidence that the `CrashActionExecutor` is broken:
```
	bool Execute(const UnsyncedAction& action) const final {
		int* a = nullptr;
		*a = 0;
		return true;
	}
```
This attempts to cause a SIGSEGV crash. However, this is actually undefined behaviour, and worse, it's UB which the compiler can trivially detect. Which means it may simply decide that this cannot happen and just return true. Or it can emit code that formats your HD. Both are valid.

If one wants to generate a crash we could simply call `std::abort`. If there exists a reason why a `SIGSEGV` is really really needed here, we can `#include <csignal>` and `raise(SIGSEGV);` (but is there really a reason to do this?)

Please do let me know which you prefer, this is 100% a bug and I'd really like to fix it.

**Comment by marcushutchings** [comment]
Since there's a standards compliant way to achieve the same result then I fully support the proposal to make the change. std::abort makes the most sense (provided the comment is updated) - @Beherith do you know whether the crash detector will catch an abort?

**Comment by Beherith** [comment]
I have not tested any other exception type, I only know that /crash works and does not require any changes. 
If you can validate via testing that any other method will also generate a correct stack trace on both win and linux platforms, then you may change it. But this has to be validated.

**Comment by Svalorzen** [comment]
I don't think I can test on Windows (not in a short amount of time anyway, I could try to setup a VM to compile everything but I'd need to figure out the env). At the same time, I can definitely check which of the two options works most closely to how it is working now. I think in the worst case manually raising SIGSEGV should in practice get identical results. I will let you updated on this.

**Comment by Svalorzen** [comment]
I'll also open a separate PR for that, I think it's a different enough change to warrant it. I'll link it here when it's ready.

**Comment by Svalorzen** [comment]
Finally finished! Now, to test it all.

- [ ] DumpStateActionExecutor
- [ ] HangActionExecutor
- [ ] LODScaleActionExecutor
- [ ] DistDrawActionExecutor
- [ ] IconFadeVanishActionExecutor
- [ ] IconFadeStartActionExecutor
- [ ] IconScaleActionExecutor
- [ ] DistIconActionExecutor
- [ ] MaxViewRangeActionExecutor
- [ ] MinViewRangeActionExecutor
- [ ] MaxNanoParticlesActionExecutor
- [ ] MaxParticlesActionExecutor
- [ ] ParticleDrawOrderActionExecutor
- [ ] ParticleSoftenActionExecutor
- [ ] DistSortProjectilesActionExecutor
- [ ] EndGraphActionExecutor
- [ ] VSyncActionExecutor
- [ ] TeamHighlightActionExecutor
- [ ] CrossActionExecutor
- [ ] GrabInputActionExecutor
- [ ] SetGamespeedActionExecutor
- [ ] GroundDetailActionExecutor
- [ ] FullscreenActionExecutor
- [ ] HardwareCursorActionExecutor
- [ ] LuaGarbageCollectControlExecutor
- [ ] SpeedControlActionExecutor
- [ ] SoundChannelEnableActionExecutor
- [ ] DebugGLActionExecutor
- [ ] DebugActionExecutor
- [ ] AllyActionExecutor
- [ ] SpecFullViewActionExecutor
- [ ] SpecTeamActionExecutor
- [ ] TeamActionExecutor
- [ ] AIControlActionExecutor
- [ ] AIKillReloadActionExecutor
- [ ] SetOverlayActionExecutor
- [ ] SetActionExecutor
- [ ] SayPrivateByPlayerIDActionExecutor
- [ ] SayPrivateActionExecutor
- [ ] FeatureDrawerTypeActionExecutor
- [ ] UnitDrawerTypeActionExecutor
- [ ] WaterActionExecutor
- [ ] GetAvailableCommands
- [ ] MapShadowPolyOffsetActionExecutor
- [ ] MapMeshDrawerActionExecutor

I'll ask questions here if I can't manage to test a command for any reason :)

**Comment by Svalorzen** [comment]
## Dodn't know how to test ##
- DumpState: how to check the output?
- LodScale: what does it do?

## Works but couldn't figure out what the effects are ##

### The argument gets read correctly, but I don't know what visible effects it has ###
- MaxParticles
- ParticleSoften
- ParticleDrawOrder
- DistSortProjectiles
- EndGraph
- DistIcon

### Other ###
- DistDraw: looks like it works, however it may have a bug: when the models go out of LOD and are replaced by a 2d sprite, a 3d "afterimage" is left in place when the LOD was triggered.

**Comment by marcushutchings** [comment]
I can answer that MaxParticles creates a soft limit on the number of particals the engine will use at once. The easiest way to see it in action would be by extremes. Rockets trails, clouds, smoke, all count towards this limit. So you could set it realliy low and spawn a ton of rocket bots, then raise the limit to see more smoke particles.

**Comment by marcushutchings** [comment]
DistIcon IIRC is the distance when icons are drawn over units - so that shouldn't be too difficult to check if I'm right.

**Comment by marcushutchings** [comment]
Some of the others look like ones that Beherith and IceXuick may be able to advise on.

**Comment by Svalorzen** [comment]
> DistIcon IIRC is the distance when icons are drawn over units - so that shouldn't be too difficult to check if I'm right.

This has just been removed (which is the reason why atm the branch has conflicts), so it's not an issue anymore.

In general tho since I only touched the parsing of the command arguments, if they get parsed correctly I'm generally inclined to say that the code is correct; and if there was a bug it was already there (especially because the logic that actually does things is usually somewhere else and the executor just triggers it). So I think that if I can correctly test the parsing that should be enough for the PR to be considered correct.

If I do find something that doesn't seem to work correctly I'll obviously report it, but it should be out of scope of this PR anyway.

**Comment by Svalorzen** [comment]
Ok, I think I finished! There's only a handful of commands I didn't test, but everything else seems to work just fine. Let me know if you need anything else before merging this :)

**Comment by lhog** [comment]
Only one question from my side.
If we found any issues with your re-implementation, would you be available to fix it within reasonable amount of time (a few days)?

**Comment by Svalorzen** [comment]
Absolutely. In the end also each command is fairly simple so if I overlooked something it would be a 30 second fix most likely.

File: rts/Game/UnsyncedGameCommands.cpp [hunk 1]
```diff
@@ -3631,7 +3629,6 @@ void UnsyncedGameCommands::AddDefaultActionExecutors()
 	AddActionExecutor(AllocActionExecutor<ChatActionExecutor>("Spec", "s:", true));
 
 	AddActionExecutor(AllocActionExecutor<TrackActionExecutor>());
-	AddActionExecutor(AllocActionExecutor<TrackOffActionExecutor>());
```
Comment (badosu) [inline rts/Game/UnsyncedGameCommands.cpp:3634]: Is this action redundant with some other one?

Comment (badosu) [reply rts/Game/UnsyncedGameCommands.cpp:3634]: Ah, I see this: `// TODO merge together with "TrackOff" to "Track 0|1", and deprecate the two old ones`

Comment (badosu) [reply rts/Game/UnsyncedGameCommands.cpp:3634]: How does the `Track` argument look like?

**Comment by lhog** [comment]
LOL I misclicked and pressed merge & rebase instead of squash & merge.
Can you make another PR with the same branch please? I'm gonna force move the repo to where it was 5 min ago

**Comment by Svalorzen** [comment]
TBH I'm not sure I know how to do this. I mainly know git commands, I'm not super familiar with the GitHub PR interface. I tried simply redoing the PR but I can't seem to manage to do it.

**Comment by badosu** [comment]
In your local directory:

```
git checkout UnsyncedGameCommandsCleanup
git checkout -b UnsyncedGameCommandsCleanup2
git push
```

**Pull request merged**